Feature/automatically remove downloaded played episodes

mrkrash opened the pull request:
I have added switch in settings section for enable/disable automatic erase of played downloaded episodes.
Erasing occurs as a result of the event `AudioProcessingState.completed` in audio player service.
I also write test for settings page but I don't know how to simulate the player event and then write a test for this new feature.

Close #112

amugofjava commented on lib/services/audio/default_audio_player_service.dart:
```diff
@@ -1053,7 +1057,11 @@ class _DefaultAudioPlayerHandler extends BaseAudioHandler with SeekHandler {
 
-        await repository.saveEpisode(storedEpisode);
+        if (settings.deleteDownloadedPlayedEpisodes) {
+          repository.deleteEpisode(storedEpisode);
```
Here you are calling the repository directly to delete the episode. Although this is similar to the save above, when you delete an episode a number of additional actions are performed such as marking the episode as played (if set in settings) and tidying up any associated transcripts. By deleting directly from the repository these steps are missed.

I think it would be better to call the `deleteDownload` in the podcast service instance that is available to the audio player service. If you take a look you can see it carries out several steps before calling `deleteEpisode` on the repository.

mrkrash replied:
Hi @amugofjava, I agree. I had not paid attention to the additional actions. I make the correction.

amugofjava commented on lib/l10n/L.dart:
```diff
@@ -447,6 +447,16 @@ class L {
+  String get settings_delete_played_label {
+    return message('settings_delete_played_label') ??
+        Intl.message(
+          'Delete downloaded played episodes',
```
Great that you found the Strings settings OK as I realised now that I have no documentation on handling translations, which I will address shortly. 

I think this string would read better as `Delete downloaded episodes once played` because it is closer in language to `Mark deleted episodes as played`. I also think it this setting should be moved up one so it is directly under `Mark deleted episodes as played` and before the SD card setting.

mrkrash replied:
ok! np.

amugofjava reviewed (changes requested):
Hi @mrkrash,

Thank you for your PR. This is the first time I have used the Github code review tool, so I hope I have done this correctly and it all makes sense.

I have made a couple of comments: one for a bug and one for a change request.

Also, as you'll see from one comment, I have completed omitted any instructions on how to handle translations. I will fix this. I am happy to handle the translations for this issue, or if you're happy to give it a go I can give you some pointers.

mrkrash commented:
If you want, give me some pointers. I could add also Italian language.

amugofjava reviewed (changes requested):
Hi @mrkrash,

Thanks for the update. The wording and setting position looks good now; however, the delete logic is not correct yet.

What we want to happen is, when playback of an episode completes we need to check the auto remove download setting and, if enabled, remove the MP3 file from the device and carry out the same processing that would occur if you manually hit the delete button.

What I can see when I test this PR, is that when playback completes the episode is removed from the database, but the local MP3 delete logic does not take place. This results in the episode being removed from the episode list when you go back into that podcast and the local MP3 file still remains.

What I would recommend is to follow the logic through when playback completes to see what Anytime does and which methods are triggered as an episode completes. Where you are currently deleting the episode from the repository is likely to be where you want to call the delete download logic.

Let me know if you need any more help with this. One thing I want to do is improve the documentation for Anytime so if there are any areas you think are poorly documented or overly complicated, please let me know.

mrkrash commented on lib/services/audio/default_audio_player_service.dart:
```diff
@@ -524,6 +525,10 @@ class DefaultAudioPlayerService extends AudioPlayerService {
     log.fine('We have completed episode ${_currentEpisode?.title}');
 
+    if (settingsService.deleteDownloadedPlayedEpisodes) {
+      await repository.deleteEpisode(_currentEpisode!);
```
ok, I understand :-)

mrkrash replied:
I made the correction: now the file is deleted from the storage but the episode and its status remain correctly in the database.
Testing it, listening to the downloaded episode, staying in the downloads screen, I noticed that when the episode is finished, the screen does not refresh automatically (just exit and re-enter and it refreshes).
Analysing the 'delete' button in the episode, I see that the episode is added to a list and then the Bloc architecture (episode_bloc) does the rest.
What would be the best way to achieve both file deletion and screen sync? Should the deletion also be delegated to episode_bloc?

amugofjava replied:
Yes, I will need to give this one a little thought. Ideally, the deletion from the service would trigger a reaction further up to cause the screen to refresh. Might need a little refactoring. Will let you know what I find.

amugofjava replied:
I woke up at 4am this morning realising what the problem is! Funny how things pop in to your head like that.

In the `episode_bloc` there is an episode listener, but I have applied a filter so it only reacts to downloaded episodes. This no longer applies, as we need to react to updates that include episodes that are no longer downloaded.

In the `_listenEpisodeEvents()` method, if you remove the filter on the listener and then try again. You should find episodes are removed from the Downloads tab as they are removed. This will need a good bit of testing to ensure there are no knock-on effects of removing the filter (especially in terms of performance), but I think this should resolve the issue.

mrkrash replied:
I did some tests by removing the filter and it works perfectly. However, I thought that instead of removing the filter, it could be hooked to another change of state, namely played.
When deleting the file, following this feature, the episode will always be played, the filter will pass and the screen will update correctly.
What do you think about?

amugofjava commented:
I have been testing this. It seems to work OK on an emulator, but I am getting issues on real devices. With the option enabled, at the end of an episode playback stalls and the episode is not deleted. I suspect a lock somewhere.

Have you have any similar issues testing this on real devices?

mrkrash commented:
No, but I can test on other device.
Test it and report again.

mrkrash commented:
Hi @amugofjava,
in the last two weeks I have tested on a physical device, my Oppo Android Phone. I have used it to listen to several podcasts of different lengths and have never encountered any uncertainties or problems.
Can I do some more tests?

amugofjava commented:
Hi @mrkrash,

Thanks for the update. I'll do some more testing here, see if I can reproduce the issue. From memory, when it happened to me I had a few episodes of Podnews Daily queued up. These are short episodes and when it finished episode one, and tried to start episode two it stalled when the auto delete option was enabled.

I'll let you know what I find. Thanks.

amugofjava commented:
Hi @mrkrash,

I have managed to reproduce the issue. What is happening is that, when the delete option is enabled it is being called at the end of every episode and not just for those that have been downloaded.

To reproduce, enable the delete option and then stream an episode rather than playing a downloaded one. At the end of the episode an exception will be raised and playback will stop.

mrkrash commented:
Hi @amugofjava,
I have added a control that proceed with deletion only if episode has download status to downloaded. When you have two. minutes, can you test (I have tested but I'm unable to reproduce this thing :-( )?

amugofjava commented:
Hi @mrkrash,

Sorry for the slow response, I've had a busy few weeks. I've given this update a quick test and for me the episodes are not deleting once played. I will dive into the code when I can to see what's going on, but giving you a heads up so you can take a look too.

Is it all working for you? If you queue up a mix of downloaded and streamed episodes, are all the episodes played successfully and do all the downloaded episodes get removed?

mrkrash commented:
Hi @amugofjava, I haven't test a mixed queue, sorry.
I'll test it and verify where is the error.

mrkrash commented:
Hi @amugofjava, I'm sorry for my late reply. I have just end test on my iPhone emulator and on my Android Phone. There was an error on a condition and now the episodes are now deleted correctly. I have rebased this PR and test it with mixed queue. In the queue, episode have a length of five minutes (I couldn't find any shorter ones).

amugofjava commented:
I'll try and get this tested soon @mrkrash - just been really busy the past few weeks. Thanks.

amugofjava commented:
Thanks @mrkrash looks good and I will merge your PR.

There are a couple of small issues, but I am happy to handle these:

1. There is an edge case when using the sleep timer and setting it to 'End of episode'. At the end of the episode playback stops, but as it's at the end the mp3 is deleted which causes an exception.
2. Needs translations for Italian & German.

mrkrash commented:
Hi @amugofjava, perfect. I will pay attention in this two last issues :-)
